pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UnusedPrivateMethodRule.java

```diff
@@ -47,23 +51,33 @@ protected Collection<String> defaultSuppressionAnnotations() {
     @Override
     public Object visit(ASTCompilationUnit file, Object param) {
         // We do three traversals:
-        // - one to find methods referenced by Junit5 MethodSource annotations
+        // - one to find methods:
+        // --- referenced by any attribute of any annotation
+        // --- with name same as a method annotated with Junit5 MethodSource if the annotation value is empty
         // - one to find the "interesting methods", ie those that may be violations
         // - another to find the possible usages. We only try to resolve
         //   method calls/method refs that may refer to a method in the
         //   first set, ie, not every call in the file.
-
-        Set<String> methodsUsedByAnnotations = file.descendants(ASTMethodDeclaration.class)
-            .crossFindBoundaries()
-            .children(ASTModifierList.class)
-            .children(ASTAnnotation.class)
-            .filter(t -> TypeTestUtil.isA("org.junit.jupiter.params.provider.MethodSource", t))
-            .toStream()
-            // Get the referenced method names… if none, use the test method name instead
-            .flatMap(a -> a.getFlatValue("value").isEmpty()
-                    ? Stream.of(a.ancestors(ASTMethodDeclaration.class).first().getName())
-                    : a.getFlatValue("value").toStream().map(mv -> (String) mv.getConstValue()))
-            .collect(Collectors.toSet());
+        Set<String> methodsUsedByAnnotations =
+            file.descendants(ASTAnnotation.class)
+                .crossFindBoundaries()
+                .toStream()
+                .flatMap(a -> Stream.concat(
+                    a.getFlatValues().toStream()
+                     .map(ASTMemberValue::getConstValue)
+                     .filter(value -> value instanceof String)
+                     .map(value -> (String) value)
+                     .filter(StringUtils::isNotEmpty),
+                    TypeTestUtil.isA("org.junit.jupiter.params.provider.MethodSource", a) && a.getFlatValue("value").isEmpty()
+                        ? Optional.ofNullable(a.getParent())
+                                .map(GenericNode::getParent)
+                                .filter(m -> m instanceof ASTMethodDeclaration)
+                                .map(m -> ((ASTMethodDeclaration) m).getName())
+                                .map(Stream::of)
+                                .orElseGet(Stream::empty)
+                        : Stream.empty())
```

Comment:
I think this would do the same and avoid the ternary and dealing with Optional (not tested):
```suggestion
					NodeStream.of(a)
				              .filter(it -> TypeTestUtil.isA("org.junit.jupiter.params.provider.MethodSource", it) 
					                              && it.getFlatValue("value").isEmpty())
					          .parents().parents()
					          .filterIs(ASTMethodDeclaration.class)
					          .toStream()
					          .map(ASTMethodDeclaration::getName))
```

Reply:
Changed as suggested. Thanks!